File: pact-jvm-server/src/main/scala/au/com/dius/pact/server/Publish.scala
package au.com.dius.pact.server

import au.com.dius.pact.core.model.{OptionalBody, Request, Response}
import com.typesafe.scalalogging.StrictLogging

import scala.collection.JavaConverters._
import java.io.{File, IOException}

import au.com.dius.pact.core.pactbroker.{PactBrokerClient, RequestFailedException}

object Publish extends StrictLogging {

  def apply(request: Request, oldState: ServerState, config: Config): Result = {
    val jsonBody = JsonUtils.parseJsonString(request.getBody.valueAsString())
    val consumer: Option[String] = getVarFromJson("consumer", jsonBody)
    val consumerVersion: Option[String] = getVarFromJson("consumerVersion", jsonBody)
    val provider: Option[String] = getVarFromJson("provider", jsonBody)
    val tags: Option[::[String]] = getListFromJson("tags", jsonBody)
    val broker: Option[String] = getBrokerUrlFromConfig(config)
    val authToken: Option[String] = getVarFromConfig(config.authToken)

    var response = new Response(500, ResponseUtils.CrossSiteHeaders.asJava)
    if (broker.isDefined) {
      if (consumer.isDefined && consumerVersion.isDefined && provider.isDefined) {
        response = publishPact(consumer.get, consumerVersion.get, provider.get, broker.get, authToken, tags)
      } else {
        def errorJson: String = "{\"error\": \"body should contain consumer, consumerVersion and provider.\"}"
        def body: OptionalBody = OptionalBody.body(errorJson.getBytes())
        response = new Response(400, ResponseUtils.CrossSiteHeaders.asJava, body)
      }
    } else {
      def errorJson: String = "{\"error\" : \"Broker url not correctly configured please run server with -b or --broker 'http://pact-broker.adomain.com' option\" }"
      def body: OptionalBody = OptionalBody.body(errorJson.getBytes())
      response = new Response(500, ResponseUtils.CrossSiteHeaders.asJava, body)
    }
    Result(response, oldState)
  }

  private def publishPact(consumer: String, consumerVersion: String, provider: String, broker: String, authToken: Option[String], tags: Option[::[String]]) = {
    val fileName: String = s"$consumer-$provider.json"
    val pact = new File(s"${System.getProperty("pact.rootDir", "target/pacts")}/$fileName")

    logger.debug("Publishing pact with following details: ")
    logger.debug("Consumer: " + consumer)
    logger.debug("ConsumerVersion: " + consumerVersion)
    logger.debug("Provider: " + provider)
    logger.debug("Pact Broker: " + broker)
    logger.debug("Tags: " + tags.getOrElse(None))

    try {
      val options = getOptions(authToken)
      val brokerClient: PactBrokerClient = new PactBrokerClient(broker, options.asJava)
      val res = brokerClient.uploadPactFile(pact, consumerVersion, tags.getOrElse(List()).asJava)
      if( res.component2() == null) {
        logger.debug("Pact succesfully shared. deleting file..")
        removePact(pact)
        new Response(200, ResponseUtils.CrossSiteHeaders.asJava, OptionalBody.body(res.component1().getBytes()))
      } else {
        new Response(500, ResponseUtils.CrossSiteHeaders.asJava, OptionalBody.body(res.component2().getLocalizedMessage.getBytes()))
      }

    } catch {
      case e: IOException => new Response(500, ResponseUtils.CrossSiteHeaders.asJava, OptionalBody.body(s"""{"error": "Got IO Exception while reading file. ${e.getMessage}"}""".getBytes()))
      case e: RequestFailedException => new Response(e.getStatus.getStatusCode, ResponseUtils.CrossSiteHeaders.asJava, OptionalBody.body(e.getBody.getBytes()))
      case t: Throwable => new Response(500, ResponseUtils.CrossSiteHeaders.asJava, OptionalBody.body(t.getMessage.getBytes()))
    }
  }

  private def getOptions(authToken: Option[String]): Map[String, _] = {
    var options: Map[String, _]= Map()
    if(authToken.isDefined) {
      options = Map("authentication" -> List("bearer",authToken.get).asJava)
    }
    options
  }



  private def removePact(file: File): Unit = {
    if (file.exists()) {
      file.delete()
    }
  }

  private def getVarFromConfig(variable: String) = {
    if (!variable.isEmpty) Some(variable)
    else None
  }

  def getBrokerUrlFromConfig(config: Config): Option[String] = {
    if (!config.broker.isEmpty && config.broker.startsWith("http")) Some(config.broker)
    else None
  }

  private def getVarFromJson(variable: String, json: Any): Option[String] = json match {
    case map: Map[AnyRef, AnyRef] => {
      if (map.contains(variable)) Some(map(variable).toString)
      else None
    }
    case _ => None
  }

  private def getListFromJson(variable: String, json: Any): Option[::[String]] = json match {
    case map: Map[AnyRef, AnyRef] => {
      if (map.contains(variable)) Some(map(variable).asInstanceOf[::[String]])
      else None
    }
    case _ => None
  }

}

File: pact-jvm-server/src/main/kotlin/au/com/dius/pact/server/Publish.kt
package au.com.dius.pact.server

import au.com.dius.pact.core.model.OptionalBody
import au.com.dius.pact.core.model.Request
import au.com.dius.pact.core.model.Response
import au.com.dius.pact.core.pactbroker.IPactBrokerClient
import au.com.dius.pact.core.pactbroker.PactBrokerClient
import au.com.dius.pact.core.pactbroker.PactBrokerClientConfig
import au.com.dius.pact.core.pactbroker.RequestFailedException
import au.com.dius.pact.core.support.Json
import au.com.dius.pact.core.support.isNotEmpty
import au.com.dius.pact.core.support.json.JsonParser
import io.github.oshai.kotlinlogging.KotlinLogging
import java.io.File
import java.io.IOException

private val logger = KotlinLogging.logger {}

object Publish {

  private val CrossSiteHeaders = mapOf("Access-Control-Allow-Origin" to listOf("*"))

  @JvmStatic
  fun apply(request: Request, oldState: ServerState, config: Config): Result {
    val jsonBody = Json.fromJson(JsonParser.parseString(request.body.valueAsString()))
    val consumer = getVarFromJson("consumer", jsonBody)
    val consumerVersion = getVarFromJson("consumerVersion", jsonBody)
    val provider = getVarFromJson("provider", jsonBody)
    val tags = getListFromJson("tags", jsonBody)
    val broker = getBrokerUrlFromConfig(config)
    val authToken = getVarFromConfig(config.authToken)

    var response = Response(500, CrossSiteHeaders.toMutableMap())
    if (broker != null) {
      if (consumer != null && consumerVersion != null && provider != null) {
        val options = getOptions(authToken)
        val brokerClient = PactBrokerClient(broker, options.toMutableMap(), PactBrokerClientConfig())
        response = publishPact(consumer, consumerVersion, provider, broker, brokerClient, tags)
      } else {
        val errorJson = "{\"error\": \"body should contain consumer, consumerVersion and provider.\"}"
        val body = OptionalBody.body(errorJson.toByteArray())
        response = Response(400, CrossSiteHeaders.toMutableMap(), body)
      }
    } else {
      val errorJson = "{\"error\" : \"Broker url not correctly configured please run server with -b or --broker 'http://pact-broker.adomain.com' option\" }"
      val body = OptionalBody.body(errorJson.toByteArray())
      response = Response(500, CrossSiteHeaders.toMutableMap(), body)
    }
    return Result(response, oldState)
  }

  fun publishPact(consumer: String, consumerVersion: String, provider: String, broker: String, brokerClient: IPactBrokerClient, tags: List<String>?): Response {
    val fileName = "$consumer-$provider.json"
    val pact = File("${System.getProperty("pact.rootDir", "target/pacts")}/$fileName")

    logger.debug { "Publishing pact with following details: " }
    logger.debug { "Consumer: $consumer" }
    logger.debug { "ConsumerVersion: $consumerVersion" }
    logger.debug { "Provider: $provider" }
    logger.debug { "Pact Broker: $broker" }
    logger.debug { "Tags: $tags" }

    return try {
      val res = brokerClient.uploadPactFile(pact, consumerVersion, tags.orEmpty())
      if (res.errorValue() == null) {
        logger.debug { "Pact successfully shared. deleting file.." }
        removePact(pact)
        Response(200, CrossSiteHeaders.toMutableMap(), OptionalBody.body(res.get()!!.toByteArray()))
      } else {
        Response(500, CrossSiteHeaders.toMutableMap(), OptionalBody.body(res.errorValue()!!.localizedMessage.toByteArray()))
      }
    } catch (e: IOException) {
      Response(500, CrossSiteHeaders.toMutableMap(), OptionalBody.body("{\"error\": \"Got IO Exception while reading file. ${e.message}\"}".toByteArray()))
    } catch (e: RequestFailedException) {
      Response(e.status, CrossSiteHeaders.toMutableMap(), OptionalBody.body(e.body?.toByteArray()))
    } catch (t: Throwable) {
      Response(500, CrossSiteHeaders.toMutableMap(), OptionalBody.body(t.message?.toByteArray()))
    }
  }

  fun getOptions(authToken: String?): Map<String, Any> {
    var options = mapOf<String, Any>()
    if (authToken != null) {
      options = mapOf("authentication" to listOf("bearer", authToken))
    }
    return options
  }

  private fun removePact(file: File) {
    if (file.exists()) {
      file.delete()
    }
  }

  fun getVarFromConfig(variable: String) =
    if (!variable.isEmpty()) variable
    else null

  fun getBrokerUrlFromConfig(config: Config) =
    if (config.broker.isNotEmpty() && config.broker.startsWith("http")) config.broker
    else null

  fun getVarFromJson(variable: String, json: Any?) = when(json) {
    is Map<*, *> -> {
      if (json.contains(variable)) json[variable].toString()
      else null
    }
    else -> null
  }

  fun getListFromJson(variable: String, json: Any?): List<String>? = when(json) {
    is Map<*, *> -> {
      if (json.contains(variable)) json[variable] as List<String>
      else null
    }
    else -> null
  }
}
